engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/result/json/JsonNodes.kt

```diff
@@ -0,0 +1,148 @@
+package graphql.nadel.enginekt.transform.result.json
+
+import graphql.nadel.NadelExecutionHints
+import graphql.nadel.enginekt.transform.query.NadelQueryPath
+import graphql.nadel.enginekt.util.AnyList
+import graphql.nadel.enginekt.util.AnyMap
+import graphql.nadel.enginekt.util.JsonMap
+import graphql.nadel.enginekt.util.foldWhileNotNull
+import java.util.Collections
+
+/**
+ * Utility class to extract data out of the given
+ */
+class JsonNodes(
+    private val data: JsonMap,
+    private val executionFlags: NadelExecutionHints,
+) {
+    private val nodes = Collections.synchronizedMap(
+        mutableMapOf<NadelQueryPath, List<JsonNode>>(),
+    )
+
+    /**
+     * Extracts the nodes at the given query selection path.
+     */
+    fun getNodesAt(queryPath: NadelQueryPath, flatten: Boolean = false): List<JsonNode> {
+        if (!executionFlags.newJsonNodeTraversal) {
+            return JsonNodeExtractor.getNodesAt(data, queryPath, flatten)
+        }
+
+        val rootNode = JsonNode(JsonNodePath.root, data)
+        return getNodesAt(rootNode, queryPath, flatten)
+    }
+
+    /**
+     * Extract the node at the given json node path.
+     */
+    fun getNodeAt(path: JsonNodePath): JsonNode? {
+        if (!executionFlags.newJsonNodeTraversal) {
+            return JsonNodeExtractor.getNodeAt(data, path)
+        }
+
+        val rootNode = JsonNode(JsonNodePath.root, data)
+        return getNodeAt(rootNode, path)
+    }
+
+    /**
+     * Extracts the nodes at the given query selection path.
+     */
+    private fun getNodesAt(rootNode: JsonNode, queryPath: NadelQueryPath, flatten: Boolean = false): List<JsonNode> {
+        var queue = listOf(rootNode)
+
+        // todo work backwards here instead of forwards
```

Comment:
I'll do this in another PR.

Want to land these changes. Not sure if it's even faster. Just an idea I had.